src/client/easy.c

```diff
@@ -0,0 +1,780 @@
+	server->cache.len = (unsigned)n_servers;
+	assert((uint64_t)server->cache.len == n_servers);
+	server->cache.cap = (unsigned)n_servers;
+	return 0;
+}
+
+static int maybeJoinCluster(struct dqlite_server *server,
+			    struct client_context *context)
+{
+	int rv;
+
+	if (findNodeInCache(&server->cache, server->local_id) != NULL) {
+		return 0;
+	}
+
+	rv = clientSendAdd(&server->proto, server->local_id, server->local_addr,
+			   context);
+	if (rv != 0) {
+		clientClose(&server->proto);
+		return 1;
+	}
+	rv = clientRecvEmpty(&server->proto, context);
+	if (rv != 0) {
+		clientClose(&server->proto);
+		return 1;
+	}
+	rv = refreshNodeStoreCache(server, context);
+	if (rv != 0) {
+		return 1;
+	}
+	return 0;
+}
+
+static int bootstrapOrJoinCluster(struct dqlite_server *server,
+				  struct client_context *context)
+{
+	struct client_node_info info;
+	int rv;
+
+	if (server->is_new && server->bootstrap) {
+		rv = openAndHandshake(&server->proto, server->local_addr,
+				      server->local_id, context);
+		if (rv != 0) {
+			return 1;
+		}
+
+		info.addr = strdupChecked(server->local_addr);
+		info.id = server->local_id;
+		info.role = DQLITE_VOTER;
+		pushNodeInfo(&server->cache, info);
+	} else {
+		rv = connectToSomeServer(server, context);
+		if (rv != 0) {
+			return 1;
+		}
+
+		rv = tryReconnectToLeader(&server->proto, context);
+		if (rv != 0) {
+			return 1;
+		}
+
+		rv = refreshNodeStoreCache(server, context);
+		if (rv != 0) {
+			return 1;
+		}
+
+		rv = maybeJoinCluster(server, context);
+		if (rv != 0) {
+			return 1;
+		}
+	}
+
+	writeLocalInfo(server);
+	writeNodeStore(server);
+	return 0;
+}
+
+static void *refreshTask(void *arg)
+{
+	struct dqlite_server *server = arg;
+	struct client_context context;
+	struct timespec ts;
+	int rv;
+
+	rv = pthread_mutex_lock(&server->mutex);
+	assert(rv == 0);
+	for (;;) {
+		rv = clock_gettime(CLOCK_REALTIME, &ts);
+		assert(rv == 0);
+		ts.tv_sec += server->refresh_period;
+		rv = pthread_cond_timedwait(&server->cond, &server->mutex, &ts);
+		if (server->shutdown) {
+			break;
+		}
+		assert(rv == 0 || rv == ETIMEDOUT);
+
+		clientContextMillis(&context, 5000);
+		if (server->proto.fd == -1) {
+			rv = connectToSomeServer(server, &context);
+			if (rv != 0) {
+				continue;
+			}
+			(void)tryReconnectToLeader(&server->proto, &context);
+			if (server->proto.fd == -1) {
+				continue;
+			}
+		}
+		rv = refreshNodeStoreCache(server, &context);
+		if (rv != 0) {
+			continue;
+		}
+		writeNodeStore(server);
+	}
+	return NULL;
+}
+
+int dqlite_server_start(dqlite_server *server)
+{
+	int info_fd;
+	int store_fd;
+	ssize_t size;
+	char *buf;
+	ssize_t n_read;
+	struct client_context context;
+	int rv;
+
+	if (server->bootstrap && server->cache.len > 0) {
+		goto err;
+	}
+
+	server->is_new = true;
+	server->dir_fd = open(server->dir_path, O_RDONLY | O_DIRECTORY);
+	if (server->dir_fd < 0) {
+		goto err;
+	}
+	info_fd = openat(server->dir_fd, "info-fresh", O_RDWR | O_CREAT, 0664);
```

Comment:
What about renaming this to `server.info` or `server-info` or `info.txt`? The `-fresh` bit feels a bit like an implementation detail.